### ISSUE TITLE
Building wheel for M1/M2 chipset

### ISSUE THREAD

**Issue by jonathan-cook235** [deepmind/lab2d#23]
The Melting Pot README states that there is no appropriate dmlab2d wheel for the Apple M1/M2 chipset. It provides an example bash script for installing and setting up the wheel, but this results in an error when trying to import dmlab2d_pybind.so, saying that the version installed was still 'x86_64' rather than 'arm64'. I'm unsure what to change about the bash script to get the right version installed.

**Comment by tkoeppe** [comment]
Could you share a link to that README and describe the commands you've run?

**Comment by jonathan-cook235** [comment]
Here is the README (https://github.com/deepmind/meltingpot/blob/main/README.md) and here is the exemplar bash script for building lab2d (https://github.com/deepmind/meltingpot/blob/main/install-dmlab2d.sh).

**Comment by tkoeppe** [comment]
I didn't write that script, but those last lines https://github.com/deepmind/meltingpot/blob/6ec136b0b034c75bc4a6164c4c9264a97fd603ba/install-dmlab2d.sh#L81-L82 look suspicious. Does the `bazel build` from just above succeed? That should result in a wheel file, which you can then install.

**Comment by saran-t** [comment]
You might need to pass `--macos_cpus=arm64` to `bazel build`.

https://bazel.build/reference/command-line-reference#flag--macos_cpus

**Comment by jonathan-cook235** [comment]
Thanks, the build does succeed (and I now pass the recommended macos_cpus arg) and I'm able to install from that. However, when I try to import dmlab2d, I get the following error relating to dmlab2d_pybind:

`File "/Users/jcook/Documents/meltingpot/notebooks/dmlab2_test.py", line 1, in <module>
    import dmlab2d
  File "/Users/jcook/anaconda3/envs/meltingpot/lib/python3.10/site-packages/dmlab2d/__init__.py", line 21, in <module>
    import dmlab2d.dmlab2d_pybind as dmlab2d_pybind
ImportError: dlopen(/Users/jcook/anaconda3/envs/meltingpot/lib/python3.10/site-packages/dmlab2d/dmlab2d_pybind.so, 0x0002): tried: '/Users/jcook/anaconda3/envs/meltingpot/lib/python3.10/site-packages/dmlab2d/dmlab2d_pybind.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/jcook/anaconda3/envs/meltingpot/lib/python3.10/site-packages/dmlab2d/dmlab2d_pybind.so' (no such file), '/Users/jcook/anaconda3/envs/meltingpot/lib/python3.10/site-packages/dmlab2d/dmlab2d_pybind.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64'))
`

**Comment by tkoeppe** [comment]
What is the resulting output file from building `//dmlab2d:dmlab2d_wheel`?

**Comment by jonathan-cook235** [comment]
`INFO: From Linking dmlab2d/dmlab2d_pybind.so:
ld: warning: -undefined dynamic_lookup may not work with chained fixups
Target //dmlab2d:dmlab2d_wheel up-to-date:
  bazel-bin/dmlab2d/dmlab2d-1.0-cp310-cp310-macosx_13_0_arm64.whl
INFO: Elapsed time: 62.800s, Critical Path: 27.83s
INFO: 332 processes: 2 internal, 330 darwin-sandbox.
INFO: Build completed successfully, 332 total actions`

**Comment by tkoeppe** [comment]
`bazel-bin/dmlab2d/dmlab2d-1.0-cp310-cp310-macosx_13_0_arm64.whl` looks like it is the right wheel -- it seems to have the correct SOABI tag? Can you pip-install _that_?

**Comment by jonathan-cook235** [comment]
Yes, that pip installs fine. It's then the import of this version that throws the error. Specifically when the `dmlab2d/__init__.py` calls `import dmlab2d.dmlab2d_pybind as dmlab2d_pybind`.

**Comment by tkoeppe** [comment]
That's quite odd. Can you double-check, maybe use `strace` to see which shared object is getting loaded, and disassemble it to see if it really contains AArch64 instructions (not x86-64 ones)?

**Comment by saran-t** [comment]
@tkoeppe There's no `strace` on macOS, and `dtruss` is non-trivial to use due to System Integrity Protection.

@jonathan-cook235 What's the output of `file /Users/jcook/anaconda3/envs/meltingpot/lib/python3.10/site-packages/dmlab2d/dmlab2d_pybind.so` ?

**Comment by jonathan-cook235** [comment]
I got: `/Users/jcook/anaconda3/envs/meltingpot/lib/python3.10/site-packages/dmlab2d/dmlab2d_pybind.so: Mach-O 64-bit dynamically linked shared library x86_64`

**Comment by tkoeppe** [comment]
(Tangent: we seem to be hardcoding the library name as .so here, https://github.com/deepmind/lab2d/blob/94e37d189aee8d309bc1dfae7227676c77636e88/build_defs.bzl#L66, when really for Mach-O it should be called .dylib... oh dear. @charlesbeattie, any thoughts on this?)

**Comment by tkoeppe** [comment]
But in any case it looks like it's definitely compiling this wrong. Just to be sure, could you inspect the .so file with `otool` or `objdump` to check if it really contains x86_64 instructions, rather than AArch64 instructions? 

I don't have a Mac to try this, but if you wanted to have a stab at it yourself, you could make a very simple toy example, just `cc_binary(name="foo.dylib", srcs=["foo.cc"], linkstatic=1, linkshared=1)` and try to compile that, `bazel build :foo.dyliib -c opt`, and see if you can manage to build an ARM binary from that with some suitable set of flags. I don't know how a bazel run on an ARM Mac would somehow end up producing an x86_64 binary...

**Comment by saran-t** [comment]
> (Tangent: we seem to be hardcoding the library name as .so here,
> 
> https://github.com/deepmind/lab2d/blob/94e37d189aee8d309bc1dfae7227676c77636e88/build_defs.bzl#L66
> 
> , when really for Mach-O it should be called .dylib... oh dear. @charlesbeattie, any thoughts on this?)

That's a Python thing, it needs `.so` for extension modules on macOS.

**Comment by jagapiou** [comment]
Sounds like you may have installed the arm64 version of bazel and not the x86_64 version. Note that https://bazel.build/install/os-x still directs people at x86_64 by default.

I used https://github.com/bazelbuild/bazel/releases/download/5.3.2/bazel-5.3.2-installer-darwin-arm64.sh

**Comment by jagapiou** [comment]
Here's my mac's output:
```
$ bazel version
WARNING: Invoking Bazel in batch mode since it is not invoked from within a workspace (below a directory having a WORKSPACE file).
Build label: 5.3.2
Build target: bazel-out/darwin_arm64-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed Oct 19 18:35:48 2022 (1666204548)
Build timestamp: 1666204548
Build timestamp as int: 1666204548
```

**Comment by jonathan-cook235** [comment]
Thanks a lot. Using this bazel version resolved that issue I think, but I now have: `ImportError: dlopen(/Users/jcook/anaconda3/envs/meltingpot/lib/python3.10/site-packages/dmlab2d/dmlab2d_pybind.so, 0x0002): symbol not found in flat namespace '_CFRelease'`

**Comment by tkoeppe** [comment]
Ah, this somehow rings a bell...  Can you add `linkopts = ["-framework CoreFoundation"],` to the module build rule https://github.com/deepmind/lab2d/blob/94e37d189aee8d309bc1dfae7227676c77636e88/dmlab2d/BUILD#L7-L19?

**Comment by jonathan-cook235** [comment]
This still results in the same error. Unless there is another step I should be taking?

**Comment by tkoeppe** [comment]
Oh... maybe this linker flag (also) needs to go on the main executable?

**Comment by jonathan-cook235** [comment]
Something very strange: when I run `dmlab2d_test.py` everything completes just fine. Previously I was running this recommended lab2d test from the Melting Pot README: 
```python
    import dmlab2d
    import dmlab2d.runfiles_helper

    lab = dmlab2d.Lab2d(dmlab2d.runfiles_helper.find(), {"levelName": "chase_eat"})
    env = dmlab2d.Environment(lab, ["WORLD.RGB"])
    env.step({})
``` 

Bizarrely, that Python code still gives me the same import error, but the `dmlab2d_test.py` file does not.

**Comment by jagapiou** [comment]
Check your PYTHONPATH, sys.path etc. Make sure `import dmlab2d` imports dmlab2d from whereever you installed the wheel to and not from your local git repo.

**Comment by jonathan-cook235** [comment]
All problems fixed, thanks a lot for all your help.

**Comment by tkoeppe** [comment]
Can you explain what was ultimately going on, just out of curiosity? :-)

**Comment by elliottower** [comment]
@jagapiou any idea about non M1/M2 macs having this same error? https://github.com/deepmind/lab2d/issues/24

**Comment by jonathan-cook235** [comment]
After I installed the correct version of Bazel, linked previously in the
thread, I continued to get the import error when running the Python code I
quoted. Setting up a new environment following the same steps, but with the
newly installed Bazel meant I was able to pass the given lab2d install
tests and run a melting pot scenario (which depends on lab2d). I'm still
not exactly sure what the route core of the import error was.
